internal/gitea/repo.go

```diff
@@ -0,0 +1,218 @@
+			fileName = path.Join(repo.folder, fileName)
+		}
+		bytes, _, err = repo.client.GetFile(
+			repo.org,
+			repo.Details.Name,
+			repo.defaultBranch,
+			fileName,
+			false,
+		)
+		if err == nil {
+			break
+		}
+	}
+	if len(bytes) == 0 {
+		return errors.New("error getting README for repo")
+	}
+
+	content := string(bytes)
+
+	// Parse contents of README to identify associated Snaps and Charms.
+	readme := &repos.Readme{Body: content}
+	ctx := context.Background()
+	repo.Details.Snap = readme.LinkedSnap(ctx)
+	repo.Details.Charm = readme.LinkedCharm(ctx)
+
+	return nil
+}
+
+// processReleases fetches a repository's releases from Gitea, then populates
+// r.Details.Releases with the information in the relevant format for releasegen.
+func (repo *Repository) processReleases() error {
+	// TODO: This currently gets all releases across the entire repository, even
+	// in a monorepo. It's not clear what happens with releases in a Gitea
+	// monorepo - for example, are they not used at all? Do they all start with
+	// the name of the 'subrepo'?
+
+	opts := gitea.ListReleasesOptions{
+		ListOptions:  gitea.ListOptions{PageSize: releasesPerRepo},
+		IsDraft:      gitea.OptionalBool(false),
+		IsPreRelease: gitea.OptionalBool(false),
+	}
+
+	source := repo.Details.Name
+	if repo.Details.Monorepo != "" {
+		source = repo.Details.Monorepo
+	}
+	releases, _, err := repo.client.ListReleases(repo.org, source, opts)
+	if err != nil {
+		return err
+	}
+
+	for _, rel := range releases {
+		repo.Details.Releases = append(repo.Details.Releases, &repos.Release{
+			ID:        rel.ID,
+			Version:   rel.TagName,
+			Timestamp: rel.PublishedAt.Unix(),
+			Title:     rel.Title,
+			Body:      renderReleaseBody(rel.Note, repo),
+			URL:       rel.URL,
+			CompareURL: fmt.Sprintf(
+				"%s/compare/%s...%s", repo.Details.URL, rel.TagName, repo.defaultBranch),
+		})
+	}
+
+	return nil
+}
+
+// processCommitsSinceRelease calculates the number of commits that have
+// occurred on the default branch of the repository since the last release, and
+// populates the information in r.Details.
+func (repo *Repository) processCommitsSinceRelease() error {
+	// TODO: This does not currently handle monorepos - see note about releases
+	// in `processReleases`. If there are releases in Gitea with a monorepo,
+	// then this function maybe needs to restrict the commits to ones where the
+	// tree overlaps - but maybe we actually want to know how many commits
+	// overall, because we want to count things like common code being adjusted.
+
+	if len(repo.Details.Releases) == 0 {
+		return errors.New("processCommitsSinceRelease must not be called without releases!")
+	}
+
+	// Add the commit delta between last release and default branch.
+	latestRelease := repo.Details.Releases[len(repo.Details.Releases)-1]
+	opts := gitea.ListCommitOptions{
+		ListOptions: gitea.ListOptions{PageSize: releasesPerRepo},
+		SHA:         latestRelease.Version,
+		Path:        "",
```

Comment:
`""` is the default (zero value) for strings, so normally you wouldn't include fields like this, unless there's some strong reason `Path` should be explicit here (in which case probably add a comment like `// not set because of such and such`). Similar below.